### ISSUE TITLE
HSV Support (HomeKit Secure Video)

### ISSUE THREAD

**Issue by jacksonjao** [kopiro/homebridge-tapo-camera#35]
Hi! Kopiro, first, thank you for the effort you put on this project, this should be verified for homebridge.

I wanted to ask you if it's possible to add HSV support, I am going to leave you a repo where they are using it. 

https://github.com/seydx/homebridge-camera-ui

**Comment by kopiro** [comment]
Hello, I'm gonna check and see if it's possible :)

**Comment by Pathaleon** [comment]
Would like to second this feature request :)

Will be a nice addition to this great plugin for the Tapo!

**Comment by xatrix78** [comment]
This will be awesome!

**Comment by Stianker1** [comment]
Totally agree! Would be a super addition to the allready great plugin. Is there any way to enable two way audio as well?

**Comment by atk49** [comment]
Thanks for your great plug-in, I totally agree this request if it’s possible.

**Comment by albinmedoc** [comment]
Is there any updates on this feature?
I'm looking forward to it 🤩

**Comment by albinmedoc** [comment]
Here are some interesting discussions about HSV

[https://github.com/homebridge/HAP-NodeJS/issues/741#issuecomment-573441082](https://github.com/homebridge/HAP-NodeJS/issues/741#issuecomment-573441082)

[https://github.com/homebridge/HAP-NodeJS/pull/920](https://github.com/homebridge/HAP-NodeJS/pull/920)

**Comment by albinmedoc** [comment]
Here are two plugins that has implemented HSV

- [homebridge-camera-ffmpeg (sub-branch)](https://github.com/Sunoo/homebridge-camera-ffmpeg/blob/hksv-old/src/recordingDelegate.ts)
- [homebridge-camera-ui](https://github.com/seydx/homebridge-camera-ui)

**Comment by kopiro** [comment]
Hello; thanks @albinmedoc for reporting these work.

It looks like camera-ffmpeg plugin is still working on it; once they release a stable version of HSV I'm going to bump the version and in theory we should get this for free.

**Comment by mivalme** [comment]
Hello @kopiro 
Already camera-ffmpeg release a version with HSV
Could you check the videoProcessor parameter that use camera-ffmpeg, I need to set that parameter from tapo-camera

Thanks, great job

**Comment by kopiro** [comment]
Hello @mivalme , where do you see the release of camera-ffmpeg with HSV?

**Comment by x23piracy** [comment]
Hi,

can someone clearify if camera ffmpeg has the hsv feature or not, afai can see there is not.
Asking because i likely purchase more tapos and i found a nice working way over scrypted but i really like to get rid of it again if a homebridge plugin can/will support it.


Best Regards X23

**Comment by kopiro** [comment]
Hello folks! Update on this: I've tried to implement it and even though I managed to get the camera tilt/pan working (https://github.com/kopiro/homebridge-tapo-camera/pull/95) - I've not been able to find an integration between Homebridge and Homekit itself that allows this functionality in Homekit app.

I don't really like the solution of having fake switches that controls that, therefore I'm not going to implement an hack.

**Comment by x23piracy** [comment]
@kopiro i fully agree the fake switches are just a dirty workaround, native support would be the best.
Thanks for testing.

**Comment by SwizzVPN** [comment]
@kopiro What about HSV/HKSV? Would it be possible to take the code from camera-ui? I really want to ditch scrypted.

**Comment by fgabrielbr** [comment]
Hello! Is it possible to add the HSV feature to this plugin?

**Comment by vfontes** [comment]
Pity HKSV hasn't been implemented, having only this plugin instead of having to mess with Scrypted would be so much better.

**Comment by kopiro** [comment]
hi, I'm not using this plug-in anymore as I am moved to Home Assistant, but if anyone wants to help, I can help with the review

**Comment by kopiro** [comment]
No, Home Assistant doesn't have a HKSV functionality either. But with HA I just don't need it as those features are integrated into HA

**Comment by kopiro** [comment]
Current status: upstream camera-ffmpeg has now released HKSV support, but it is only available through the renamed `@homebridge-plugins/homebridge-camera-ffmpeg` v4 package.

Adopting that package here is not a dependency-only update: v4 is ESM and requires Node 20+, while this plugin is currently CommonJS and supports Node 18. It also changes the streaming delegate constructor and recording configuration. Shipping it in the automatic patch-release flow would therefore break a supported runtime.

Keeping this enhancement open and marking it `help wanted`. A safe implementation needs to be treated as a deliberate compatibility migration—either with a new major version that drops Node 18 and moves to the v4 API, or after the plugin's support floor otherwise advances. It should wire the existing Tapo ONVIF motion service into HKSV event recording and be validated with real camera/HomeKit hardware before release.

**Comment by kopiro** [comment]
An experimental implementation is now available in draft PR #226.

The PR includes source-install instructions, the opt-in configuration, and a checklist for reporting camera/Home hub compatibility. Please keep feedback on the PR and include sanitized debug/FFmpeg logs. This remains a user-testing build and will not be merged until real HKSV clips and existing live streaming have been validated on representative hardware.

**Comment by kopiro** [comment]
We are looking for users willing to test the experimental HKSV implementation in draft PR #226 on real Tapo cameras.

Please follow the source-install and configuration instructions in the PR, then report your camera model/hardware/firmware, Node and Homebridge versions, Apple Home hub, whether existing snapshots/live streaming still work, and whether motion produces playable HKSV clips. Testing both with and without `enableHKSVPrebuffer` would be especially useful.

Please post results and sanitized debug/FFmpeg logs directly on #226 so all compatibility findings stay together. Do not use this build if you cannot comfortably restore your existing plugin version; it is an unmerged v3 test build and drops Node 18 support.
